### PR TITLE
[SPARK-14676] Wrap and re-throw Await.result exceptions in order to capture full stacktrace

### PR DESCRIPTION
**Pull request by JoshRosen**
When `Await.result` throws an exception which originated from a different thread, the resulting stacktrace doesn't include the path leading to the `Await.result` call itself, making it difficult to identify the impact of these exceptions. For example, I've seen cases where broadcast cleaning errors propagate to the main thread and crash it but the resulting stacktrace doesn't include any of the main thread's code, making it difficult to pinpoint which exception crashed that thread.

This patch addresses this issue by explicitly catching, wrapping, and re-throwing exceptions that are thrown by `Await.result`.

I tested this manually using https://github.com/JoshRosen/spark/commit/16b31c825197ee31a50214c6ba3c1df08148f403, a patch which reproduces an issue where an RPC exception which occurs while unpersisting RDDs manages to crash the main thread without any useful stacktrace, and verified that informative, full stacktraces were generated after applying the fix in this PR.

/cc @rxin @nongli @yhuai @anabranch

### CODE REVIEW

**Comment by JoshRosen** [comment]
Hmm, those are legitimate test failures. The problem was that wrapping something in `Exception` changed the exception matched by some test asserts.

**Comment by JoshRosen** [comment]
I've decided to just leave the occurrence in `FutureAction` untouched for now.

**Comment by JoshRosen** [comment]
Hmm, this is unfortunate. It looks like there are a bunch more assertions like the one in the `FutureAction` test suite which match on the class / name of the exception thrown from methods which call `Await.result`. I wonder whether this change of wrapping the exception is going to cause regressions in user code which tries to do the same thing.

Should I try to use reflection to make a best-effort attempt to re-throw using the same exception?

File: core/src/main/scala/org/apache/spark/network/BlockTransferService.scala [hunk 1]
```diff
@@ -101,7 +102,12 @@ abstract class BlockTransferService extends ShuffleClient with Closeable with Lo
         }
       })
 
-    Await.result(result.future, Duration.Inf)
+    try {
+      Await.result(result.future, Duration.Inf)
+    } catch {
+      case NonFatal(t) =>
+        throw new Exception("Exception occurred while fetching block", t)
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/network/BlockTransferService.scala]: RuntimeException? or SparkException.

**Comment by rxin** [comment]
Maybe instead of doing this, we should create a wrapper function that wraps around await.result?

**Comment by thunterdb** [comment]
+1, because we can then add a linter rule that disallows `Await.result`

**Comment by JoshRosen** [comment]
Alright, I've updated this to add a `ThreadUtils.awaitResult` and added a linter rule to ban direct use of `Await.result`.

**Comment by rxin** [comment]
cc @sameeragarwal for review

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 2]
```diff
@@ -260,7 +260,12 @@ private[spark] class BlockManager(
   def waitForAsyncReregister(): Unit = {
     val task = asyncReregisterTask
     if (task != null) {
-      Await.ready(task, Duration.Inf)
+      try {
+        Await.ready(task, Duration.Inf)
```
Comment (ScrapCodes) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Unrelated to this PR, But waiting for infinite time has a downside, that if this (main)thread blocks then the app running will appear to have hanged with no way to know unless one checks the thread dump somehow. However if it is for finite time duration, an exception is thrown on timeout. In the case `Duration.Inf` there is no exception ever thrown.

If I am correct about the above, I am not sure why it is being used widely ? I am just asking so I understand if there is some side to it that I do not understand.

Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: According to the Scaladoc (and actual usages), it looks like this particular `waitForAsyncReregister` method is only used in test code and I'm guessing that it's probably called from within an interrupt-based timeout block.

As for the other usages, we'd have to consider them on a case-by-case basis.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 3]
```diff
@@ -802,7 +807,12 @@ private[spark] class BlockManager(
       logDebug("Put block %s locally took %s".format(blockId, Utils.getUsedTimeMs(startTimeMs)))
       if (level.replication > 1) {
         // Wait for asynchronous replication to finish
-        Await.ready(replicationFuture, Duration.Inf)
+        try {
+          Await.ready(replicationFuture, Duration.Inf)
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: @ScrapCodes, towards your other comment, I think that timeouts in this case might already happen to be covered by network / RPC timeouts within the `replicationFuture`'s code.

**Comment by sameeragarwal** [comment]
LGTM

**Comment by rxin** [comment]
Merging in master. Thanks.

**Pull request closed without merging**